lib/jobs/get-catalog-values-job.js

```diff
@@ -0,0 +1,85 @@
+'use strict';
+
+var di = require('di');
+
+module.exports = getCatalogValuesJobFactory;
+di.annotate(getCatalogValuesJobFactory, new di.Provide('Job.Get.Catalog.Values'));
+di.annotate(getCatalogValuesJobFactory, new di.Inject(
+    'Job.Base',
+    'Logger',
+    'Services.Waterline',
+    'Util',
+    '_'
+));
+
+function getCatalogValuesJobFactory(
+    BaseJob,
+    Logger,
+    waterline,
+    util,
+    _
+){
+    var logger = Logger.initialize(getCatalogValuesJobFactory);
+
+    /**
+     * This job finds one or more requested values from one or more specified catalogs.
+     * The catalog values are made available in the shared context in an object that 
+     * maps the catalog values to requester provided keys.
+     *
+     * Example requestedData array:
+     * [{
+     *   "source": "ohai",
+     *   "keys": {
+     *     "myIPAddrKeyName": "data.ipaddress"
+     *   }
+     * }]
+     *
+     * Adds to context.data:
+     * {
+     *   "myIPAddrKeyName": <catalog value>
+     * }
+     *
+     * @param options
+     * @param context
+     * @param taskId
+     * @constructor
+     */
+
+    function GetCatalogValuesJob(options, context, taskId){
+        GetCatalogValuesJob.super_.call(this, logger, options, context, taskId);
+        this.nodeId = context.target;
+        this.options = options;
+        this.logger = logger;
+    }
+
+    util.inherits(GetCatalogValuesJob, BaseJob);
+
+    /**
+     * @memberOf GetCatalogValuesJob
+     */
+    GetCatalogValuesJob.prototype._run = function _run(){
+        var self = this,
+            newObj = {};
+
+        // Get the correct catalog based on source
+        _.forEach(self.options.requestedData, function(catalogInfo){
```

Comment:
@dmskinner : Actually, this is a problem about how to control the timing sequence using Promise, I will use following example to describe what's the problem:
```javascript
var _ = require('lodash');
var Promise= require('bluebird');

var requestedData = ['a', 'b', 'c'];
var context = {};
_.forEach(requestedData, function(data) {
    Promise.resolve(data)
    .then(function(val) {
        return val + '1';
    })
    .then(function(val) {
        return val + '2';
    })
    .then(function(val) {
        context[data] = val;
        console.log("call _done!");
    })
    .catch(function(err) {
        console.log("call _done(err)!");
    });
});

console.log('context #1:');
console.log(context);

setTimeout(function() {
    console.log('context #2');
    console.log(context);
}, 2000);
```

The output is
```
context #1:
{}
call _done!
call _done!
call _done!
context #2
{ a: 'a12', b: 'b12', c: 'c12' }
```

The problems are:
- The _done is called thrice, actually we only allow it is called after all catalogs are handled.
- The context has no data when the _.forEach finishes, because the javascript code is async.